File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.*;
import java.util.*;

public class Main {

    static class Point {
        int y, x;

        Point(int y, int x) {
            this.y = y;
            this.x = x;
        }

    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        // String temp = br.readLine();
        // StringTokenizer st = new StringTokenizer(br.readLine());
        // int N = Integer.parseInt(st.nextToken());
        // int M = Integer.parseInt(st.nextToken());

        int N = Integer.parseInt(br.readLine());

        int height = 0;
        int answer = 0;
        int arr[][] = new int[N][N];
        Queue<Point> q = new LinkedList<>();
        Queue<Point> q_2 = new LinkedList<>();

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int r = 0; r < N; r++) {
                int temp = Integer.parseInt(st.nextToken());
                arr[i][r] = temp;
                if (temp > height) {
                    height = temp;
                }
            }
        }
        int d_y[] = { 0, 0, -1, 1 };
        int d_x[] = { -1, 1, 0, 0 };
        // N = y
        // M = x
        while (height > -1) {
            boolean visited[][] = new boolean[N][N];
            for (int i = 0; i < N; i++) {
                for (int r = 0; r < N; r++) {

                    if (arr[i][r] > height - 1) {
                        q.add(new Point(i, r));
                    }
                }
            }
            int cnt = 0;
            while (!q.isEmpty()) {
                Point point = q.poll();

                if (!visited[point.y][point.x]) {
                    visited[point.y][point.x] = true;
                    q_2.add(new Point(point.y, point.x));
                    cnt++;
                    while (!q_2.isEmpty()) {
                        Point search = q_2.poll();
                        for (int i = 0; i < 4; i++) {
                            int n_y = search.y + d_y[i];
                            int n_x = search.x + d_x[i];

                            if (n_y >= 0 && n_y < N && n_x >= 0 && n_x < N) {
                                if (!visited[n_y][n_x] && arr[n_y][n_x] > height - 1) {
                                    visited[n_y][n_x] = true;
                                    q_2.add(new Point(n_y, n_x));
                                }
                            }
                        }
                    }
                }
            }
            if (cnt > answer) {
                answer = cnt;
            }
            height--;
        }
        System.out.println(answer);
    }
}

File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <bits/stdc++.h>
using namespace std;

struct Point
{
    int y, x;
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, height = 0, temp, ret = 0;

    cin >> N;
    vector<vector<int>> arr(N, vector<int>(N, 0));

    for (int i = 0; i < N; i++)
    {
        for (int r = 0; r < N; r++)
        {
            cin >> temp;
            arr[i][r] = temp;
            if (temp > height)
            {
                height = temp;
            }
        }
    }

    while (height > -1)
    {
        vector<vector<bool>> visited(N, vector<bool>(N, false));
        queue<Point> q;
        for (int i = 0; i < N; i++)
        {
            for (int r = 0; r < N; r++)
            {
                if (arr[i][r] > height)
                {
                    q.push({i, r});
                }
            }
        }

        int d_y[4] = {-1, 1, 0, 0};
        int d_x[4] = {0, 0, -1, 1};

        int t_ret = 0;
        queue<Point> q_2;

        while (!q.empty())
        {
            Point qq = q.front();
            q.pop();
            if (!visited[qq.y][qq.x])
            {
                visited[qq.y][qq.x] = true;
                t_ret++;
                q_2.push({qq.y, qq.x});

                while (!q_2.empty())
                {
                    Point search = q_2.front();
                    q_2.pop();
                    for (int e = 0; e < 4; e++)
                    {
                        int n_y = search.y + d_y[e];
                        int n_x = search.x + d_x[e];

                        if (n_y >= 0 && n_y < N && n_x >= 0 && n_x < N)
                        {
                            if (!visited[n_y][n_x] && arr[n_y][n_x] > height)
                            {
                                visited[n_y][n_x] = true;
                                q_2.push({n_y, n_x});
                            }
                        }
                    }
                }
            }
        }

        if(ret < t_ret){
            ret = t_ret;
        }
        height--;
    }

    cout << ret;

    return 0;
}
